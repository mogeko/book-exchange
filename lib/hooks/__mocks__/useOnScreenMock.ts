import * as hooks from "@/lib/hooks/useOnScreen";

const mock = jest.spyOn(hooks, "default");

const useOnScreenMock = {
  target: hooks.default,
  visiable: () => mock.mockReturnValue(true),
  not: {
    visiable: () => mock.mockReturnValue(false),
  },
};

export default useOnScreenMock;
