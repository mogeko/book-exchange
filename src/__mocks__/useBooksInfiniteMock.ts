import * as hooks from "@/lib/hooks/useBooks";
import { faker } from "@faker-js/faker";

const mock = jest.spyOn(hooks, "useBooksInfinite");
const setSizeMock = jest.fn();

const useBooksInfiniteMock = {
  target: hooks.useBooksInfinite,
  returnResult: (pages = 3) => ({
    success: () =>
      mock.mockImplementation((param) =>
        genExampleRes({}, [pages, param!.limit!])
      ),
    error: () =>
      mock.mockImplementation((param) =>
        genExampleRes({ data: undefined, isError: true }, [
          pages,
          param!.limit!,
        ])
      ),
    loading: () =>
      mock.mockImplementation((param) =>
        genExampleRes({ data: undefined, isLoading: true }, [
          pages,
          param!.limit!,
        ])
      ),
  }),
  genExampleRes,
  setSizeMock,
};

export const exampleData: BookTypes = {
  title: faker.word.noun(20),
  cover: faker.image.image(1280, 1910),
  description: faker.lorem.paragraph(10),
  published: faker.date.past().toISOString(),
  publisher: faker.company.companyName(),
  tags: Array.from({ length: faker.datatype.number({ min: 2, max: 8 }) }, () =>
    faker.lorem.words(2)
  ),
  author: faker.name.firstName(),
  rates: faker.datatype.number(100),
  id: faker.datatype.number(100),
};

function genExampleRes(res: Partial<ResType> = {}, [x, y] = [3, 10]): ResType {
  const exampleRes = {
    data: Array.from({ length: x }, () =>
      Array.from({ length: y }, () => exampleData)
    ),
    isValidating: false,
    isError: false,
    isLoading: false,
    mutate: jest.fn(),
    setSize: setSizeMock,
    size: 1,
  };

  return { ...exampleRes, ...res };
}

type ResType = ReturnType<typeof hooks.useBooksInfinite>;
type BookTypes = hooks.BookTypes;

export default useBooksInfiniteMock;
