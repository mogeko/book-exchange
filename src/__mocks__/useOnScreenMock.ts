import * as hooks from "@/lib/useOnScreen";

const useOnScreenMock = {
  target: hooks.default,
  visiable: () => jest.spyOn(hooks, "default").mockImplementation(() => true),
  not: {
    visiable: () =>
      jest.spyOn(hooks, "default").mockImplementation(() => false),
  },
};

export default useOnScreenMock;
