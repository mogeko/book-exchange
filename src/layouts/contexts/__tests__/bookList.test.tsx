import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import useBooksInfiniteMock from "@/__mocks__/useBooksInfiniteMock";
import useOnScreenMock from "@/__mocks__/useOnScreenMock";
import BookList from "@/layouts/contexts/bookList";

describe("bookList", () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it("renders a BookList", () => {
    useOnScreenMock.not.visiable();
    useBooksInfiniteMock.returnResult(4, 7).success();
    const { container } = render(<BookList limit={10} offset={20} />);
    const mockTarget = useBooksInfiniteMock.target;

    expect(mockTarget).toBeCalledWith({ limit: 10, offset: 20 });
    expect(container.querySelectorAll("figure").length).toBe(28);
  });

  it("snapshot a BookList", () => {
    useOnScreenMock.not.visiable();
    useBooksInfiniteMock.returnResult().success();
    const { container } = render(<BookList limit={10} offset={20} />);

    expect(container).toMatchSnapshot();
  });

  it("renders a BookList when the screen touches the bottom", () => {
    useOnScreenMock.visiable();
    useBooksInfiniteMock.returnResult().success();
    render(<BookList limit={10} offset={20} />);

    expect(useBooksInfiniteMock.setSizeMock).toBeCalledTimes(1);
    expect(useBooksInfiniteMock.setSizeMock).toBeCalledWith(2);
  });

  it("renders a BookList then error", () => {
    useOnScreenMock.not.visiable();
    useBooksInfiniteMock.returnResult().error();
    render(<BookList limit={10} offset={20} />);

    expect(screen.getByText("Network Error!")).toBeInTheDocument();
  });

  it("snapshot a BookList then error", () => {
    useOnScreenMock.not.visiable();
    useBooksInfiniteMock.returnResult().error();
    const { container } = render(<BookList limit={10} offset={20} />);

    expect(container).toMatchSnapshot();
  });

  it("renders a Display BookList then loading", () => {
    useOnScreenMock.not.visiable();
    useBooksInfiniteMock.returnResult().loading();
    const { container } = render(<BookList limit={10} offset={20} />);

    expect(container.querySelector("div.animate-pulse")).toBeInTheDocument();
  });

  it("snapshot a BookList then loading", () => {
    useOnScreenMock.not.visiable();
    useBooksInfiniteMock.returnResult().loading();
    const { container } = render(<BookList limit={10} offset={20} />);

    expect(container).toMatchSnapshot();
  });
});
