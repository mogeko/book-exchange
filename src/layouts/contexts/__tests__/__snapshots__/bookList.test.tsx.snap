// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookList with abnormal state snapshot a BookList then error 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div
      class="alert alert-error shadow-lg"
    >
      <div>
        <svg
          class="w-6 h-6"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M8.6 1c1.6.1 3.1.9 4.2 2 1.3 1.4 2 3.1 2 5.1 0 1.6-.6 3.1-1.6 4.4-1 1.2-2.4 2.1-4 2.4-1.6.3-3.2.1-4.6-.7-1.4-.8-2.5-2-3.1-3.5C.9 9.2.8 7.5 1.3 6c.5-1.6 1.4-2.9 2.8-3.8C5.4 1.3 7 .9 8.6 1zm.5 12.9c1.3-.3 2.5-1 3.4-2.1.8-1.1 1.3-2.4 1.2-3.8 0-1.6-.6-3.2-1.7-4.3-1-1-2.2-1.6-3.6-1.7-1.3-.1-2.7.2-3.8 1-1.1.8-1.9 1.9-2.3 3.3-.4 1.3-.4 2.7.2 4 .6 1.3 1.5 2.3 2.7 3 1.2.7 2.6.9 3.9.6zM7.9 7.5L10.3 5l.7.7-2.4 2.5 2.4 2.5-.7.7-2.4-2.5-2.4 2.5-.7-.7 2.4-2.5-2.4-2.5.7-.7 2.4 2.5z"
            fill-rule="evenodd"
          />
        </svg>
        <span>
          Network Error!
        </span>
      </div>
    </div>
    <div />
  </div>
  <div />
</div>
`;

exports[`BookList with abnormal state snapshot a BookList then loading 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div
      class="relative flex mb-4 w-full gap-3 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="flex flex-col w-full gap-2"
      >
        <div
          class="mb-2 w-1/5 h-2 bg-slate-700 rounded"
        />
        <div
          class="mb-2 w-1/6 h-2 bg-slate-700 rounded"
        />
        <div>
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/6 h-2 bg-slate-700 rounded"
          />
        </div>
      </div>
    </div>
    <div
      class="relative flex mb-4 w-full gap-3 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="flex flex-col w-full gap-2"
      >
        <div
          class="mb-2 w-1/5 h-2 bg-slate-700 rounded"
        />
        <div
          class="mb-2 w-1/6 h-2 bg-slate-700 rounded"
        />
        <div>
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/6 h-2 bg-slate-700 rounded"
          />
        </div>
      </div>
    </div>
    <div
      class="relative flex mb-4 w-full gap-3 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="flex flex-col w-full gap-2"
      >
        <div
          class="mb-2 w-1/5 h-2 bg-slate-700 rounded"
        />
        <div
          class="mb-2 w-1/6 h-2 bg-slate-700 rounded"
        />
        <div>
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/6 h-2 bg-slate-700 rounded"
          />
        </div>
      </div>
    </div>
    <div
      class="relative flex mb-4 w-full gap-3 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="flex flex-col w-full gap-2"
      >
        <div
          class="mb-2 w-1/5 h-2 bg-slate-700 rounded"
        />
        <div
          class="mb-2 w-1/6 h-2 bg-slate-700 rounded"
        />
        <div>
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/6 h-2 bg-slate-700 rounded"
          />
        </div>
      </div>
    </div>
    <div
      class="relative flex mb-4 w-full gap-3 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="flex flex-col w-full gap-2"
      >
        <div
          class="mb-2 w-1/5 h-2 bg-slate-700 rounded"
        />
        <div
          class="mb-2 w-1/6 h-2 bg-slate-700 rounded"
        />
        <div>
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/6 h-2 bg-slate-700 rounded"
          />
        </div>
      </div>
    </div>
    <div
      class="relative flex mb-4 w-full gap-3 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="flex flex-col w-full gap-2"
      >
        <div
          class="mb-2 w-1/5 h-2 bg-slate-700 rounded"
        />
        <div
          class="mb-2 w-1/6 h-2 bg-slate-700 rounded"
        />
        <div>
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/6 h-2 bg-slate-700 rounded"
          />
        </div>
      </div>
    </div>
    <div
      class="relative flex mb-4 w-full gap-3 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="flex flex-col w-full gap-2"
      >
        <div
          class="mb-2 w-1/5 h-2 bg-slate-700 rounded"
        />
        <div
          class="mb-2 w-1/6 h-2 bg-slate-700 rounded"
        />
        <div>
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/6 h-2 bg-slate-700 rounded"
          />
        </div>
      </div>
    </div>
    <div
      class="relative flex mb-4 w-full gap-3 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="flex flex-col w-full gap-2"
      >
        <div
          class="mb-2 w-1/5 h-2 bg-slate-700 rounded"
        />
        <div
          class="mb-2 w-1/6 h-2 bg-slate-700 rounded"
        />
        <div>
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/6 h-2 bg-slate-700 rounded"
          />
        </div>
      </div>
    </div>
    <div
      class="relative flex mb-4 w-full gap-3 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="flex flex-col w-full gap-2"
      >
        <div
          class="mb-2 w-1/5 h-2 bg-slate-700 rounded"
        />
        <div
          class="mb-2 w-1/6 h-2 bg-slate-700 rounded"
        />
        <div>
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/6 h-2 bg-slate-700 rounded"
          />
        </div>
      </div>
    </div>
    <div
      class="relative flex mb-4 w-full gap-3 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="flex flex-col w-full gap-2"
      >
        <div
          class="mb-2 w-1/5 h-2 bg-slate-700 rounded"
        />
        <div
          class="mb-2 w-1/6 h-2 bg-slate-700 rounded"
        />
        <div>
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/5 h-2 bg-slate-700 rounded"
          />
          <div
            class="mb-2 w-4/6 h-2 bg-slate-700 rounded"
          />
        </div>
      </div>
    </div>
  </div>
  <div />
</div>
`;

exports[`bookList snapshot a BookList 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div />
  </div>
  <div />
</div>
`;
